SpicesUpdater@claudiux/files/SpicesUpdater@claudiux/applet.js

```diff
@@ -0,0 +1,1529 @@
+        fstream.write(chunk.get_data(), null);
+      }
+    }));
+
+    let ret;
+    // Queue of the http request
+    _httpSession.queue_message(request, Lang.bind(this, function(session, message) {
+      // close the file
+      fstream.close(null);
+    }));
+    let triplet = new Array(uuid, filePath, destdir);
+    this.ready_to_unzip[type].push(triplet)
+  }; // End of download_zip
+  */
+
+  unzip(zipname, dir) {
+    log("!!!!! unziping "+ zipname + " in " + dir);
+    let [success, pid] = GLib.spawn_async(null,
+                                          ['unzip', '-uod', dir, '--', zipname],
+                                          null,
+                                          GLib.SpawnFlags.SEARCH_PATH | GLib.SpawnFlags.DO_NOT_REAP_CHILD,
+                                          null);
+    return success;
+  }; // End of unzip
+
+  is_to_check(type) {
+    return (this.types_to_check.indexOf(type) > -1);
+  }; // End of is_to_check
+
+  is_to_update(type) {
+    return (this.types_to_update.indexOf(type) > -1);
+  }; // End of is_to_update
+
+  get_can_be_updated(type) {
+    var spicesList;
+    switch (type) {
+      case "applets":
+        spicesList = this.unprotectedAppletsList;
+        break;
+      case "themes":
+        spicesList = this.unprotectedThemesList;
+        break;
+      case "extensions":
+        spicesList = this.unprotectedExtensionsList;
+        break;
+      case "desklets":
+        spicesList = this.unprotectedDeskletsList;
+
+    }
+    var ret = [];
+    for (let s of spicesList) {
+      if (s["isunprotected"] === true) {
+        ret.push(s["name"])
+      }
+    }
+    log("can_be_updated("+ type + ") = " + JSON.stringify(ret));
+    return ret
+  }; // End of get_can_be_updated
+
+  get_must_be_updated(type) {
+    let can_be_updated = this.get_can_be_updated(type);
+    var ret = new Array();
+    var lc, lm;
+    for (let uuid of can_be_updated) {
+      lc = this._get_last_edited_from_cache(type, uuid);
+      if (lc !== null) {
+        lm = this._get_last_edited_from_metadata(type, uuid);
+        if (lm !== null) {
+          if (lc > lm) ret.push(uuid);
+        }
+      }
+    }
+    //log("!!! get_must_be_updated("+ type + ") = " + ret);
+    return ret
+  }; // End of get_must_be_updated
+
+  get_active_spices(type) {
+    // Returns the list of active spices of type 'type'
+    var dconfEnabled;
+    var elt = 0;
+    switch (type) {
+      case "applets":
+        dconfEnabled = "/org/cinnamon/enabled-applets";
+        elt = 3;
+        break;
+      case "themes":
+        dconfEnabled = "/org/cinnamon/theme/name";
+        break;
+      case "extensions":
+        dconfEnabled = "/org/cinnamon/enabled-extensions";
+        break;
+      case "desklets":
+        dconfEnabled = "/org/cinnamon/enabled-desklets";
+
+    }
+    let [res, out, err, status] = GLib.spawn_command_line_sync('dconf read "'+dconfEnabled+'"');
```

Comment:
I think it would be better to make these spawn calls asynchronous.

Reply:
I replaced calls to `dconf` via `GLib.spawn_command_line_sync` with calls to `Gio.Settings`.
(Near line 1283.)
```
_interface_settings = new Gio.Settings({ schema_id: _SETTINGS_SCHEMA });
 enabled = _interface_settings.get_string(_SETTINGS_KEY);
```

Reply:
Now, the applet tests if metadata.json is present, possibly in the subfolder /cinnamon for themes, before opening it.

Reply:
I do not understand why the widget is not on the stage, and how to put it on. (Line 509 of applet.js.)